modular_citadel/code/modules/fish/fishtank.dm

```diff
@@ -0,0 +1,676 @@
+		visible_message("A new [fish_type.fish_name] has hatched in \the [src]!")
+	//Null type fish are dud eggs, give a message to inform the player
+	else
+		to_chat(usr, "The eggs disolve in the water. They were duds!")
+
+/obj/machinery/fishtank/proc/harvest_eggs(var/mob/user)
+	if(!egg_count)									//Can't harvest non-existant eggs
+		return
+
+	if(egg_count > max_fish)						//Make sure the number of eggs doesn't exceed the max_fish for the tank
+		egg_count = max_fish						//If you somehow exceeded the cap, set the egg_count to max, destroy the excess later
+
+	while(egg_count > 0)							//Loop until you've harvested all the eggs
+		var/obj/item/fish_eggs/egg = pick(egg_list)	//Select an egg at random
+		egg = new egg(get_turf(user))				//Spawn the egg at the user's feet
+		egg_list.Remove(egg)						//Remove the egg from the egg_list
+		egg_count --								//Decrease the egg_count and begin again
+
+	egg_list.Cut()									//Destroy any excess eggs, clearing the egg_list
+
+/obj/machinery/fishtank/proc/harvest_fish(var/mob/user)
+	if(!fish_count)									//Can't catch non-existant fish!
+		to_chat(usr, "There are no fish in \the [src] to catch!")
+		return
+	var/list/fish_names_list = list()
+	for(var/datum/fish/fish_type in fish_list)
+		fish_names_list += list("[fish_type.fish_name]" = fish_type)
+	var/caught_fish = input("Select a fish to catch.", "Fishing") as null|anything in fish_names_list		//Select a fish from the tank
+	if(caught_fish)
+		user.visible_message("[user.name] harvests \a [caught_fish] from \the [src].", "You scoop \a [caught_fish] out of \the [src].")
+		var/datum/fish/fish_type = fish_names_list[caught_fish]
+		var/fish_item = fish_type.fish_item
+		if(fish_item)
+			new fish_item(get_turf(user))			//Spawn the appropriate fish_item at the user's feet.
+		kill_fish(fish_type)						//Kill the caught fish from the tank
+
+/obj/machinery/fishtank/proc/destroy(var/deconstruct = 0)
+	var/turf/open/T = get_turf(loc)									//Store the tank's turf for atmos updating after deletion of tank
+	if(!deconstruct)												//Check if we are deconstructing or breaking the tank
+		var/shards_left = shard_count
+		while(shards_left > 0)										//Produce the appropriate number of glass shards
+			new /obj/item/shard(get_turf(src))
+			shards_left --
+		if(water_level)												//Spill water that was left in the tank when it broke
+			if(istype(T))
+				T.MakeSlippery(TURF_WET_WATER)
+
+	else															//We are deconstructing, make glass sheets instead of shards
+		var/sheets = shard_count + 1								//Deconstructing it salvages all the glass used to build the tank
+		new /obj/item/stack/sheet/glass(get_turf(src), sheets)		//Produce the appropriate number of glass sheets, in a single stack
+	qdel(src)														//qdel the tank and it's contents
+	T.air_update_turf(1)											//Update the air for the turf, to avoid permanent atmos sealing with wall tanks
+
+/* /obj/machinery/fishtank/proc/spill_water()
```

Comment:
i'd normally say don't leave in commented code unless you absolutely have to but we're a downstream so it matters less.